lnst/Recipes/ENRT/ConfigMixins/FirewallMixin.py

```diff
@@ -0,0 +1,155 @@
+from abc import ABC, abstractmethod
+from lnst.Recipes.ENRT.ConfigMixins import BaseSubConfigMixin
+from lnst.RecipeCommon.FirewallControl import FirewallControl
+import copy
+
+class FirewallMixin(BaseSubConfigMixin, ABC):
+    """
+    A config mixin to apply custom firewall rulesets on hosts.
+    Do not inherit directly, use one of the derived classes below instead.
+    """
+
+    _fwctl = {}
+
+    def fwctl(self, host):
+        try:
+            return self._fwctl[host]
+        except KeyError:
+            self._fwctl[host] = host.init_class(FirewallControl)
+            return self._fwctl[host]
+
+    @property
+    def firewall_rulesets(self):
+        """
+        This property holds a dictionary of firewall rulesets in textual
+        representation, indexed by the host it should be applied to. A typical
+        use is:
+
+        { self.matched.host1: <host1 ruleset>,
+          self.matched.host2: <host2 ruleset> }
+
+        This property is used by the default `firewall_rules_generator()`
+        method. Overwriting the latter from a recipe constitutes an alternative
+        to implementing it.
+
+        The rulesets will be applied by a derived class's _apply_ruleset()
+        method, i.e. typically fed into 'nft -f' or 'iptables-restore'.
+        """
+        return {}
+
+    @firewall_rulesets.setter
+    def firewall_rulesets(self, rulesets):
+        """
+        This setter is called with all hosts' rulesets after each test run.
+        Overwrite it to perform post processing or analysis on contained state.
+        """
+        pass
+
+    @property
+    def firewall_rulesets_generator(self):
+        """
+        A generator yielding { host: ruleset, ... } to apply for a test run.
+        Each yield will turn into a new sub configuration and thus be tested
+        separately.
+        """
+        return [self.firewall_rulesets]
+
+    def generate_sub_configurations(self, config):
```

Comment:
one more thing here, this method needs to call `super()` like this:
```python
def generate_sub_configurations(self, config):
    for parent_config in super().generate_sub_configurations(config):
        ...
        # the rest of the method but now using parent_config instead of config
```
As seen in other config mixins, e.g. https://github.com/LNST-project/lnst/blob/master/lnst/Recipes/ENRT/ConfigMixins/OffloadSubConfigMixin.py#L31C1-L31C74

the `generate_sub_configurations` method is in ENRT package designed as a [collaborative inheritance](https://medium.com/swlh/cooperative-multiple-inheritance-paradigm-in-python-f048b7ecdb29) method, this is part of the "ConfigMixin" design.

This is specific to the ENRT test set (so lnst normally doesn't use this it's just this specific test set). And the collaborative inheritance is a python speciality...

basically:

```python
SomeRecipe(ConfigMixin1, ConfigMixin2, BaseEnrtRecipe):
    pass

SomeRecipe2(ConfigMixin2, ConfigMixin1, BaseEnrtRecipe):
    pass
```

Means that if both config mixin classes define the `generate_sub_configurations` methods, the **collaborate** on the result via inheritance. The call order of the collaboration depends on the class definition and what the order of base classes was PLUS the way that you use the `super()` call in your collaborative methods.

In short - `super` calls "the next in line" method, not "strict parent" method and when every definition of the method in the MRO correctly uses the `super` call, they together pass around the same `config` object and each method adds its own part to it.

If one definition of the method in the class inheritance hierarchy DOESN'T call `super` correctly it interrupts the chain of "next in line" and the design stops working. This may be intentional in some cases but in this one it's not wanted.

Reply:
Ah, indeed. Thanks for the details. I don't think colaborative inheritance is my problem, as I do call the super() object's method. But I missed the fact that it is a generator and thus will possibly return multiple different sub configs and I should mix into all of them.